internal/pkg/agent/application/upgrade/step_unpack.go

```diff
@@ -64,34 +117,46 @@ func unzip(log *logger.Logger, archivePath, dataDir string) (string, error) {
 			}
 		}()
 
-		//get hash
 		fileName := strings.TrimPrefix(f.Name, fileNamePrefix)
 		if fileName == agentCommitFile {
-			hashBytes, err := io.ReadAll(rc)
-			if err != nil || len(hashBytes) < hashLen {
-				return err
-			}
-
-			hash = string(hashBytes[:hashLen])
+			// we already loaded the hash, skip this one
 			return nil
 		}
 
+		mappedPackagePath := pm.Map(fileName)
+
 		// skip everything outside data/
-		if !strings.HasPrefix(fileName, "data/") {
+		if !strings.HasPrefix(mappedPackagePath, "data/") {
 			return nil
 		}
 
-		path := filepath.Join(dataDir, strings.TrimPrefix(fileName, "data/"))
+		dstPath := strings.TrimPrefix(mappedPackagePath, "data/")
+		dstPath = filepath.Join(dataDir, dstPath)
 
 		if f.FileInfo().IsDir() {
-			log.Debugw("Unpacking directory", "archive", "zip", "file.path", path)
+			log.Debugw("Unpacking directory", "archive", "zip", "file.path", dstPath)
 			// remove any world permissions from the directory
-			_ = os.MkdirAll(path, f.Mode()&0770)
+			_, err = os.Stat(dstPath)
+			if errors.Is(err, fs.ErrNotExist) {
+				if err := os.MkdirAll(dstPath, f.Mode().Perm()&0770); err != nil {
+					return fmt.Errorf("creating directory %q: %w", dstPath, err)
+				}
+			} else if err != nil {
+				return fmt.Errorf("stat() directory %q: %w", dstPath, err)
+			} else {
+				// set the appropriate permissions
+				err = os.Chmod(dstPath, f.Mode().Perm()&0o770)
```

Comment:
Typo?

```suggestion
				err = os.Chmod(dstPath, f.Mode().Perm()&0770)
```

Reply:
Also looks like we could calculate `f.Mode().Perm()&0770)` once and just reuse in each of the following lines.

Reply:
That is consistency with the old code, I tried to be consistent in using `0o770` format in my changes. Not sure if there's a guideline about one or the other